### PR TITLE
I3S-Picking - Prepare for removing pickFeatures flag

### PR DESCRIPTION
**Pull request by dryabinin-actionengine**
1. Add `useFeatureIds` option to the I3S loader for switching between `features` and `instance` picking mode on the` loaders.gl `side.
2. Done preparation for removing useless `pickFeatures` flag from Mesh Layer in terms of [deck.gl PR](https://github.com/visgl/deck.gl/pull/5757)
3. Add tests
3. Done code cleanup
4. Updated docs

### CODE REVIEW

File: modules/i3s/docs/api-reference/i3s-loader.md [hunk 1]
```diff
@@ -168,6 +168,7 @@ const visibleTiles = tileset3d.tiles.filter(tile => tile.selected);
...
 | `options.i3s.useDracoGeometry`      | `Bool`           | `true`  | Use 'Draco' compressed geometry to show if applicable                                                                                                                                                                                                                                                                           |
 | `options.i3s.useCompressedTextures` | `Bool`           | `true`  | Use "Compressed textures" (_.dds or _.ktx) if available and supported by GPU                                                                                                                                                                                                                                                    |
+| `options.i3s.useFeatureIds`         | `Bool`           | `true`  | Add `featureIds` property in tile content if available in geometry. `featureIds` specified in [tiles-3d-loader](https://loaders.gl/modules/3d-tiles/docs/api-reference/tiles-3d-loader) .                                                                                                                                       |
```
Comment (ibgreen) [inline modules/i3s/docs/api-reference/i3s-loader.md]: We should specify this in the `tiles` module (the common format for I3S and 3D Tiles) and link to that. We should not cross reference I3S and 3D Tiles.

Please land https://github.com/visgl/loaders.gl/pull/1408 first and then rebase and update this.

File: modules/i3s/test/lib/parsers/parse-i3s-tile-content.spec.js [hunk 2]
```diff
@@ -95,7 +95,8 @@ test('ParseI3sTileContent#should have featureIds', async t => {
   const data = await response.arrayBuffer();
   const result = await parseI3STileContent(data, tile, i3sTilesetData, {
     i3s: {
-      useDracoGeometry: false
+      useDracoGeometry: false,
```
Comment (ibgreen) [inline modules/i3s/test/lib/parsers/parse-i3s-tile-content.spec.js:98]: Can we turn draco back on?

**Review by ibgreen — COMMENTED** [review]
What do we gain by adding a `useFeatureIds`? 

Would we not always include these in the output if available?

If it controls loading, then maybe call it `loadFeatureIds`?

File: modules/i3s/test/lib/parsers/parse-i3s-tile-content.spec.js (hunk 2 above)
Comment (dryabinin-actionengine) [reply modules/i3s/test/lib/parsers/parse-i3s-tile-content.spec.js]: Done

**Comment by dryabinin-actionengine** [comment]
> What do we gain by adding a `useFeatureIds`?
> 
> Would we not always include these in the output if available?
> 
> If it controls loading, then maybe call it `loadFeatureIds`?

The idea of such option is to avoid creating one more `pickFeatures` flag in `deck.gl` to control `instance` or `feature` picking functionality. If we want to use picking by features we will include them in the tile content and if MeshLayer will detect featureIds in tile content it automatically switch `feature` picking mode is ON. If there are no any `featureIds` in the tile content `instance` picking mode will be activated.

File: modules/i3s/docs/api-reference/i3s-loader.md (hunk 1 above)
Comment (dryabinin-actionengine) [reply modules/i3s/docs/api-reference/i3s-loader.md]: Ok, will start from #1408

**Comment by ibgreen** [comment]
Hmm, gave this another thought:

I think a Tile3DLayer user may still want to dynamically control picking in deck.gl to either be per tile or per feature (we certainly do this in the i3s debug application), so I am not convinced that we should not add any new prop in deck.gl. 

Having to add remove the featureIds in the tiles to control the type of picking doesn't sound convenient. I'd vote for always loading the feature ids.

In the future, I would love to improve deck.gl picking so that it can pick multiple values at the same time - this is limited because we are currently rendering picking information into an 8 bit RGBA buffer but this is a legacy restriction, we could e.g render into a RGBA32 buffer (or use multidraw to render into multiple buffers).

@Pessimistress

**Comment by Pessimistress** [comment]
This is not what I suggested. If we can identify individual features in a tile, then there is no reason to pick per tile, so

- loaders should always load object ids if available
- deck should pick feature if object ids are present
- deck falls back to pick tile otherwise

**Comment by ibgreen** [comment]
We'll still need to handle switching on tile based picking in loaders.gl then. Ideally we can do it without traversing all our tiles and removing the object ids.

**Comment by Pessimistress** [comment]
I'm confused. Picking will identify the source tile regardless of whether feature picking is available. Don't you already have color per tile for debug? What does picking per tile do that you can't achieve otherwise?

**Comment by ibgreen** [comment]
No worries - I am the confused one. We do get the tile back, this is not an instanced situation. 

What is the default here? The difference would be between vertex index picking and feature index picking? And this is what we don't need a flag for?

**Comment by ibgreen** [comment]
Isn't his ultimately the same issue with have with the BitmapLayer
- We lose the ability to pick the instance when introducing some other kind of picking. 
- Since both the BitmapLayer and MeshLayer are often used with just one instance, removing instanced picking is often OK. 
- As far as I can tell, the 8.5 bitmap layer completely removes support for instanced picking
- But in the MeshLayer case we would maintain it depending on what attributes are provided?

That seems a bit subtle? Would it be better to just support object picking and not instanced picking until we improve the picking module (similar to the BitmapLayer)? I assume we can set a constant object Id attribute 0.

**Comment by Pessimistress** [comment]
If there is no feature index then all vertices will use the same picking color.

**Comment by ibgreen** [comment]
Whatever picking options we chose to offer (or not) in deck.gl (I am fine with either approach), I am not convinced it makes sense to me to add a loader flag for including attributes or not. 

If there is another reason for not including attributes (they take a lot of space, they are slow to load etc) I could see this flag make sense.

But if the reason is to control picking it seems too indirect - What if we wanted to change the picking mode? We would need to reload the data?

Since we have only one mesh per MeshLayer we still get both the tile and the featureId from the updated deck.gl picking process so we don't really need this flag?

**Comment by ibgreen** [comment]
A note is that this is not just for picking - it is also relates to controlling highlighting (per tile or per object). So we still have to add and remove the objectIds in the application. Works for now but we may want to come back to it.

**Pull request closed without merging**